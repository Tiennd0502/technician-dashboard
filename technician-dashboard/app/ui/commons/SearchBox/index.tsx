'use client';

import { memo, useState, useCallback, ChangeEvent, KeyboardEvent } from 'react';
import { SearchIcon } from '@chakra-ui/icons';
import { Input, InputGroup, InputRightElement } from '@chakra-ui/react';
import { removeSpaces } from '@/lib/utils';

interface SearchBoxProps {
  onSearch: (value: string) => void;
}

const SearchBox = ({ onSearch }: SearchBoxProps) => {
  const [keyword, setKeyword] = useState<string>('');

  const handleSearch = useCallback(() => onSearch(keyword || ''), [keyword, onSearch]);

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => setKeyword(removeSpaces(event?.target.value || '')),
    [],
  );

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLInputElement>) => {
      if (event.key === 'Enter') {
        onSearch(keyword);
      }
    },
    [keyword, onSearch],
  );

  return (
    <InputGroup>
      <Input
        placeholder='Enter keyword...'
        value={keyword}
        onChange={handleChange}
        onBlur={handleSearch}
        fontSize='sm'
        onKeyDown={handleKeyDown}
      />
      <InputRightElement cursor='pointer' data-testid='search-icon'>
        <SearchIcon onClick={handleSearch} />
      </InputRightElement>
    </InputGroup>
  );
};

export default memo(SearchBox);
