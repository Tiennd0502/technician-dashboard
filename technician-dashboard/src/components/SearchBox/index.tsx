import { memo, useState, useCallback, ChangeEvent, KeyboardEvent } from 'react';
import { SearchIcon } from '@chakra-ui/icons';
import { Input, InputGroup, InputRightElement } from '@chakra-ui/react';

interface SearchBoxProps {
  onSearch: (value: string) => void;
}

const SearchBox = ({ onSearch }: SearchBoxProps) => {
  const [keyword, setKeyword] = useState<string>('');

  const handleSearch = useCallback(() => onSearch(keyword || ''), [keyword, onSearch]);

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) =>
      setKeyword(event?.target.value.replace(/\s+/g, ' ') || ''),
    [],
  );

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLInputElement>) => {
      if (event.key === 'Enter') {
        onSearch(keyword);
      }
    },
    [keyword, onSearch],
  );

  return (
    <InputGroup>
      <Input
        placeholder='Enter keyword...'
        value={keyword}
        onChange={handleChange}
        onBlur={handleSearch}
        fontSize='sm'
        onKeyDown={handleKeyDown}
      />
      <InputRightElement cursor='pointer' data-testid='search-icon'>
        <SearchIcon onClick={handleSearch} />
      </InputRightElement>
    </InputGroup>
  );
};

export default memo(SearchBox);
