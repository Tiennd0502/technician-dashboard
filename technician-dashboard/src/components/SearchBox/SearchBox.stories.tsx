import type { Meta, StoryObj } from '@storybook/react';

import SearchBox from '.';

const meta: Meta<typeof SearchBox> = {
  title: 'Components/SearchBox',
  component: SearchBox,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof SearchBox>;

export const Default: Story = {
  args: { onSearch: () => null },
};
