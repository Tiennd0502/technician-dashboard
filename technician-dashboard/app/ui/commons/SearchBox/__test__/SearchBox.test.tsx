import { fireEvent, render } from '@testing-library/react';
import SearchBox from '..';

const onSearch = jest.fn();

describe('SearchBox component', () => {
  const props = {
    onSearch,
  };

  test('should render correctly', () => {
    const { container } = render(<SearchBox {...props} />);

    expect(container).toMatchSnapshot();
  });

  test('should call the search handler when blur input', () => {
    const { getByPlaceholderText } = render(<SearchBox {...props} />);
    const searchInput = getByPlaceholderText('Enter keyword...');

    fireEvent.change(searchInput, { target: { value: 'keyword' } });
    fireEvent.blur(searchInput);

    expect(onSearch).toHaveBeenCalledWith('keyword');
  });

  test('should show search icon', () => {
    const { getByTestId } = render(<SearchBox {...props} />);
    const target = getByTestId('search-icon');

    expect(target).toBeTruthy();
  });
});
