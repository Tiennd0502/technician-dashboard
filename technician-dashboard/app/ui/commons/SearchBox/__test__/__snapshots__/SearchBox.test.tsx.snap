// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox component should render correctly 1`] = `
<div>
  <div
    class="chakra-input__group css-bx0blc"
    data-group="true"
  >
    <input
      class="chakra-input css-g5hezf"
      placeholder="Enter keyword..."
      value=""
    />
    <div
      class="chakra-input__right-element css-eezmgu"
      data-testid="search-icon"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
